File: SysOp/GP.java
package SysOp;

import SysOp.Sistema.Word;

public class GP {
    PCB[] processos;
    Word[] mem;
    private GM gm;

    public GP(Word[] m, int tamPag){
        processos = new PCB[m.length/tamPag]; //Não sei o tamanho
        for(int i = 0; i < processos.length; i++)
            processos[i] = null;
        gm = new GM(m.length, tamPag);
        mem = m;}

    boolean criaProcesso(Word[] prog){
        int[] tabPag = null;
        if(!gm.aloca(prog.length, tabPag)) return false; //verifica se tem espaço na memoria
        for(int i = 0; i < processos.length; i++) //coloca na lista de processos
            if(processos[i] == null){
                processos[i] = new PCB(i, tabPag);
                break;}
        return true;}

    void desalocaProcesso(int id){
        gm.desaloca(processos[id].tabPag);
        processos[id] = null;  //aparentemente não existe um delete
    }
}

File: SysOp/cpp/gp.cpp
#include "gp.hpp"

using namespace std;

GP::GP(vector<Word>& m, int tamP) : tamPag(tamP), processosProntos(m.size() / tamP, nullptr), mem(m), gm(m.size(),tamP){}

bool GP::criaProcesso(vector<Word>& prog){
    vector<int> tabPag;
    if(!gm.aloca(prog.size(), tabPag)){ return false;}

    int numPag = static_cast<int>(tabPag.size());
    for (int p = 0; p < numPag; ++p) {
        for (int j = 0; j < tamPag; ++j){
        mem[tabPag[p] * tamPag + j] = prog[p * tamPag + j];
            
        }
    }

    for (size_t i = 0; i < processosProntos.size(); ++i){
        if(processosProntos[i] == nullptr){
            processosProntos[i] = new PCB(static_cast<int>(i), tabPag);
            break;
        }
    }

    return true;
}

void GP::desalocaProcesso(int id){
    if(id >= 0 && id < static_cast<int>(processosProntos.size()) && processosProntos[id] != nullptr){
        gm.desaloca(processosProntos[id]->tabPag);
        delete processosProntos[id];
        processosProntos[id];
        processosProntos[id] = nullptr;
    }
}

GP::~GP(){
    for(auto proc : processosProntos){
        delete proc;
    }
}

